force_wfmanager/left_side_pane/new_entity_modal.py

```diff
@@ -0,0 +1,106 @@
+from traitsui.list_str_adapter import ListStrAdapter
+
+from force_bdss.api import (BaseMCOModel, BaseMultiCriteriaOptimizerBundle,
+                            BaseDataSourceModel, BaseDataSourceBundle,
+                            BaseKPICalculatorModel, BaseKPICalculatorBundle)
+
+from .workflow_model_view import WorkflowModelView
+from .view_utils import get_bundle_name
+
+
+class ListAdapter(ListStrAdapter):
+    """ Adapter for the list of available MCOs/Data sources/KPI calculators
+    bundles """
+    def get_text(self, object, trait, index):
+        return get_bundle_name(self.item)
+
+
+class ModalHandler(Handler):
+    def object_add_button_changed(self, info):
+        if info.object.current_model is not None:
+            info.object.workflow.add_entity(info.object.current_model)
+        info.ui.dispose(True)
+        info.object.clear_model()
+
+    def object_cancel_button_changed(self, info):
+        info.ui.dispose(False)
+        info.object.clear_model()
+
+
+class NewEntityModal(HasStrictTraits):
+    """ Dialog which allows the user to add a new MCO/Data Source/KPI
+    calculator to the workflow """
+    workflow = Instance(WorkflowModelView)
+
+    #: Available bundles
+    available_bundles = Either(
+        List(Instance(BaseMultiCriteriaOptimizerBundle)),
+        List(Instance(BaseDataSourceBundle)),
+        List(Instance(BaseKPICalculatorBundle)),
+    )
+
+    #: Selected bundle in the list
+    selected_bundle = Either(
+        Instance(BaseMultiCriteriaOptimizerBundle),
+        Instance(BaseDataSourceBundle),
+        Instance(BaseKPICalculatorBundle)
+    )
+
+    add_button = Button("Add")
+    cancel_button = Button('Cancel')
+
+    #: Currently editable model
+    current_model = Either(
+        Instance(BaseMCOModel),
+        Instance(BaseDataSourceModel),
+        Instance(BaseKPICalculatorModel)
+    )
+
+    _models = Dict()
+
+    traits_view = View(
+        VGroup(
+            HSplit(
+                UItem(
+                    "available_bundles",
+                    editor=ListStrEditor(
+                        adapter=ListAdapter(),
+                        selected="selected_bundle"),
+                ),
+                UItem('current_model', style='custom', editor=InstanceEditor())
+            ),
+            HGroup(
+                UItem(
+                    'add_button',
+                    enabled_when="selected_bundle is not None"
+                ),
+                UItem('cancel_button')
+            )
+        ),
+        title='New Element',
+        handler=ModalHandler(),
+        width=800,
+        height=600,
+        kind="livemodal"
+    )
+
+    def clear_model(self):
+        self.selected_bundle = None
+        self.current_model = None
+        self._models = {}
+
+    @on_trait_change("selected_bundle")
+    def update_current_model(self):
+        if self.selected_bundle is None:
+            return
```

Comment:
also remove the model from the view, otherwise you keep showing the model when you have no selection.